### PR TITLE
defi security review rust

### PR DESCRIPTION
**Pull request by Bownairo**
Resolve all Rust comments from security review, and fix tests on latest dfx version.

### CODE REVIEW

File: rust/defi/src/defi_dapp/exchange.rs [hunk 1]
```diff
@@ -136,33 +107,45 @@ impl Exchange {
         to_amount: Nat,
     ) -> OrderPlacementReceipt {
         ic_cdk::println!("place order");
+        if from_amount <= 0u8 || to_amount <= 0u8 {
```
Comment (jplevyak) [inline rust/defi/src/defi_dapp/exchange.rs:110]: Why 0u8 here and 0u16 elsewhere?

File: rust/defi/src/defi_dapp/exchange.rs [hunk 2]
```diff
@@ -176,83 +159,69 @@ impl Exchange {
...
-                            if ((BigUint::from(b_to_amount) * BigUint::from(b.from_amount))
-                                % BigUint::from(b.to_amount))
-                                != BigUint::from(0u32)
-                            {
-                                continue;
-                            }
-                        }
-                    }
-                }
+            let mut a_to_amount = 0u16.into();
```
Comment (jplevyak) [inline rust/defi/src/defi_dapp/exchange.rs:192]: I would prefer to see the types here.

File: rust/defi/src/defi_dapp/exchange.rs [hunk 3]
```diff
@@ -176,83 +159,69 @@ impl Exchange {
...
+
+        Ok(())
     }
 
-    fn process_trade(&mut self, a: OrderId, b: OrderId, a_to_amount: u128, b_to_amount: u128) {
+    fn process_trade(
+        &mut self,
+        a: OrderId,
+        b: OrderId,
+        a_to_amount: Nat,
```
Comment (jplevyak) [inline rust/defi/src/defi_dapp/exchange.rs:222]: Just for my edification: what is the advantage of Nat over BigUint?

File: rust/defi/src/defi_dapp/exchange.rs [hunk 4]
```diff
@@ -263,63 +232,67 @@ impl Exchange {
...
-        );
-        balances.add_balance(&order_b.owner, &order_b.to_token_canister_id, b_to_amount);
+        balances.subtract_balance(&order_b.owner, &order_b.from, b_from_amount.clone());
+        balances.add_balance(&order_b.owner, &order_b.to, b_to_amount.clone());
 
         // The DEX keeps any tokens not required to satisfy the parties.
         let dex_amount_a = a_from_amount - b_to_amount;
-        if dex_amount_a > 0 {
-            balances.add_balance(&ic_cdk::id(), &order_a.from_token_canister_id, dex_amount_a);
+        if dex_amount_a > 0u16 {
```
Comment (jplevyak) [inline rust/defi/src/defi_dapp/exchange.rs:256]: Not a huge fan of implicit coercion in sensitive code.

**Review by jplevyak — APPROVED** [review]
Approved, pending closing out my comments.  Thanx!

File: rust/defi/src/defi_dapp/exchange.rs [hunk 5]
```diff
@@ -60,21 +33,20 @@ impl BalancesState {
...
     ) -> bool {
         if let Some(balances) = self.0.get_mut(owner) {
             if let Some(x) = balances.get_mut(token_canister_id) {
-                match (*x).checked_sub(delta) {
-                    Some(num) => {
-                        *x = num;
+                if *x >= delta {
+                    *x -= delta;
+                    // no need to keep an empty token record
+                    if *x == 0u16 {
```
[the remark this answers is not in the document]
Comment (Bownairo) [reply rust/defi/src/defi_dapp/exchange.rs]: Fixed!

File: rust/defi/src/defi_dapp/exchange.rs (hunk 1 above)
Comment (Bownairo) [reply rust/defi/src/defi_dapp/exchange.rs]: Just a lack of consistency, this has all been replaced with explicit construction.

File: rust/defi/src/defi_dapp/exchange.rs (hunk 3 above)
Comment (Bownairo) [reply rust/defi/src/defi_dapp/exchange.rs:232]: More-so, security was worried about conversion to `u128` during calls overflowing. I've removed this conversion altogether, and while making this change I used `Nat` for inputs, outputs, and internal types just to keep it consistent. `Nat` is backed with `BigUint` anyway.

**Pull request merged**